lib/internal/modules/esm/loader.js

```diff
@@ -338,25 +338,36 @@ class ESMLoader {
    * would have a cache key of https://example.com/foo and baseURL
    * of https://example.com/bar
    *
-   * MUST BE SYNCHRONOUS for import.meta initialization
-   * MUST BE CALLED AFTER receiving the url body due to I/O
-   * @param {string} url
-   * @returns {string}
+   * ! MUST BE SYNCHRONOUS for import.meta initialization
+   * ! MUST BE CALLED AFTER receiving the url body due to I/O
+   * @param {URL['href']} url
+   * @returns {string|Promise<URL['href']>}
    */
   getBaseURL(url) {
     if (
       StringPrototypeStartsWith(url, 'http:') ||
       StringPrototypeStartsWith(url, 'https:')
     ) {
-      // The request & response have already settled, so they are in
-      // fetchModule's cache, in which case, fetchModule returns
+      // When using network-imports, the request & response have already settled
+      // so they are in fetchModule's cache, in which case, fetchModule returns
       // immediately and synchronously
-      url = fetchModule(new URL(url), { parentURL: url }).resolvedHREF;
-      // This should only occur if the module hasn't been fetched yet
-      if (typeof url !== 'string') { // [2]
-        throw new ERR_INTERNAL_ASSERTION(`Base url for module ${url} not loaded.`);
+      // BUT, when a custom loader is used, the cache is bypassed and it hits an
+      // unsettled promise.
+      const module = fetchModule(new URL(url), { parentURL: url });
+
+      if (typeof module?.resolvedHREF === 'string') { // [2]
+        url = module.resolvedHREF;
+      } else { // This should only occur if the module hasn't been fetched yet
+        if (getOptionValue('--experimental-network-imports')) {
+          throw new ERR_INTERNAL_ASSERTION(
+            `Base url for module ${url} not loaded.`
+          );
+        }
+        // A custom loader was used instead of network-imports
+        url = module; // ! this is an unsettled promise
       }
     }
+
```

Comment:
One issue here is that using a custom https loader even with experimental network imports being unused will still result in an unused fetch request being created from Node.js core. We don't really need this extra request though at all. Here's a suggested alternative approach to avoid re-fetching network URLs when not using experimental network imports:


```js
if (getOptionValue('--experimental-network-imports') && (StringPrototypeStartsWith(url, 'http:') || StringPrototypeStartsWith(url, 'https:') {
  const cached = fetchModuleCache.get(url);
  if (cached) {
    // existing current logic here
  }
  // URL was not previously requested through the fetch cache - just use the identity non-remapped form
  else { 
    return url;
  }
}
```